### ISSUE TITLE
Implement ducc backend to pyshtools gravity and magnetic classes

### ISSUE THREAD

**Issue by spicydonkey** [SHTOOLS/SHTOOLS#454]
The spherical harmonic computation backends for the gravity and magnetic models are only implemented with the default Fortran `shtools`.  This PR proposes implementing the ducc backend to the gravity and magnetic model classes which would provide a (1) speedup to methods such as `expand`, and (2) handling of higher spherical harmonic degrees and orders.

**Comment by mreineck** [comment]
Doing this in the same fashion as with the standard spherical harmonic transforms will not work, I think. This is because the harmonic operations for gravity and magnetic models in `SHTOOLS` are implemented as special-purpose Fortran functions doing exacty the requested job. `ducc` doesn't have these particular functions, so just switching the backend will not do the trick.
However it should be possible to reformulate the expressions for gravitational and magnetic fields in a way that uses standard SHTs with a few (presumably cheap) additional operations before and after. These could most likely be written in Python. But it's definitely not a small change ...

**Comment by MarkWieczorek** [comment]
I agree that the fortran shtools routines are slow in comparison to ducc0, and I would like to fix this at some point. However, I see two possible complications.

The first is that for the gravity and magnetic potential, the spherical harmonic coefficients are multiplied by (R/r)**l (where l is degree). Here is the equation for the gravitational potential:

<img width="399" alt="Screenshot 2024-03-18 at 11 06 37" src="https://github.com/SHTOOLS/SHTOOLS/assets/8697285/6c403e6e-f38d-4895-97d3-01231bb01201">

and here it is for the magnetic potential:

<img width="298" alt="Screenshot 2024-03-18 at 11 20 04" src="https://github.com/SHTOOLS/SHTOOLS/assets/8697285/a990db4b-0137-4cb1-8571-c0343326f0b2">

There is no real difference in the two: the two communities just use different conventions.

When doing the transform, the multiplicative prefactor will be different for the radial and horizontal components, so there is no simply way to use the existing ducc0 routines to account for this. 

Second, the shtools program allows you to compute the field on a flattened ellipsoid. This is kind of important when working with terrestrial gravity and magnetic data. So in addition to needing to multiply the coefficients by a new degree dependent prefactor, this prefactor will change with latitude.

It would almost certainly require a dedicated routine in ducc0 to do the gravity/magnetic field expansions. I suspect that it probably wouldn't be too "hard" to change ducc0 to do this, but this would have to be done upstream in the ducc0 code and not here in pyshtools.

I don't know @mreineck feels about this. It would be useful for geophysics, but probably no one else. If ducc0 did want to implement such a routine, the easiest would be to implement a single routine using orthonormalized harmonics that allowed as input the semimajor axis and flattening. We could easily preprocess the magnetic and gravity coefficients to put them into the correct normalization. Also, there is no reason to add the normal_gravity functionality, as this is easily done by post-processing the output (and we will soon change how this is done in pyshtools anyways).
